test/unit/org/apache/cassandra/utils/CassandraGenerators.java

```diff
@@ -183,29 +214,304 @@ public static Gen<String> sstableFormatNames()
+            this.keyspaceNameGen = i -> keyspace;
+            return this;
+        }
+
+        public AbstractReplicationStrategyBuilder withUserAllowed()
+        {
+            List<Strategy> allowed = Stream.of(Strategy.values()).filter(s -> s.userAllowed).collect(Collectors.toList());
+            strategyGen = SourceDSL.arbitrary().pick(allowed);
+            return this;
+        }
+
+        public AbstractReplicationStrategyBuilder withRf(Gen<Integer> rfGen)
+        {
+            this.rfGen = rfGen;
+            return this;
+        }
+
+        public AbstractReplicationStrategyBuilder withRf(int rf)
+        {
+            this.rfGen = i -> rf;
+            return this;
+        }
+
+        public AbstractReplicationStrategyBuilder withDatacenters(Gen<List<String>> networkTopologyDCGen)
+        {
+            this.networkTopologyDCGen = networkTopologyDCGen;
+            return this;
+        }
+
+        public AbstractReplicationStrategyBuilder withDatacenters(String first, String... rest)
+        {
+            if (rest.length == 0)
+            {
+                this.networkTopologyDCGen = i -> Collections.singletonList(first);
+            }
+            else
+            {
+                List<String> all = new ArrayList<>(rest.length + 1);
+                all.add(first);
+                all.addAll(Arrays.asList(rest));
+                this.networkTopologyDCGen = i -> all;
+            }
+            return this;
+        }
+
+        public Gen<AbstractReplicationStrategy> build()
+        {
+            return rs -> {
+                Strategy strategy = strategyGen.generate(rs);
+                switch (strategy)
+                {
+                    case Simple:
+                        return new SimpleStrategy(keyspaceNameGen.generate(rs),
+                                                  ImmutableMap.of(SimpleStrategy.REPLICATION_FACTOR, rfGen.generate(rs).toString()));
+                    case NetworkTopology:
+                        ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+                        List<String> names = networkTopologyDCGen.generate(rs);
+                        for (String name : names)
+                            builder.put(name, rfGen.generate(rs).toString());
+                        ImmutableMap<String, String> map = builder.build();
+                        return new TestableNetworkTopologyStrategy(keyspaceNameGen.generate(rs), map);
+                    case Meta:
+                        return new MetaStrategy(keyspaceNameGen.generate(rs), ImmutableMap.of());
+                    case Local:
+                        return new LocalStrategy(keyspaceNameGen.generate(rs), ImmutableMap.of());
+                    default:
+                        throw new UnsupportedOperationException(strategy.name());
+                }
+            };
+        }
+    }
+
+    public static class TestableNetworkTopologyStrategy extends NetworkTopologyStrategy
```

Comment:
`NetworkTopologyStrategy` validates the datacenters match a config, but that makes it a bit harder to use in some contexts... so this one avoids validating datacenters!